File: clients/android/sample/src/androidTest/kotlin/dev/arizona/sample/MultiE2ETest.kt
package dev.arizona.sample

import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onAllNodesWithText
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import org.junit.Rule
import org.junit.Test

/**
 * On-device multiple independent counters: navigate to /native/multi and tap the
 * middle counter's "+"; only its count changes. Run with
 * ./gradlew :sample:connectedCheck (server at 10.0.2.2:4040).
 */
class MultiE2ETest {
    @get:Rule
    val rule = createAndroidComposeRule<MainActivity>()

    @Test
    fun incrementsOneCounterIndependently() {
        rule.waitForText("Multi")
        rule.onNodeWithText("Multi").performClick()
        rule.waitForText("A: 0")

        // Tap counter B's "+" (the second of the three "+" buttons).
        rule.onAllNodesWithText("+")[1].performClick()
        rule.waitForText("B: 1")
        // A and C untouched.
        rule.onNodeWithText("A: 0").assertExists()
        rule.onNodeWithText("C: 0").assertExists()
    }
}

File: clients/ios/Sample/UITests/MultiE2ETest.swift
import XCTest

/// On-device multiple independent counters: navigate to /native/multi and tap the
/// middle counter's "+"; only its count changes. The Android analogue is
/// MultiE2ETest.
final class MultiE2ETest: XCTestCase {
    private var app: XCUIApplication!

    override func setUpWithError() throws {
        continueAfterFailure = false
        app = launchSample()
    }

    func testIncrementsOneCounterIndependently() {
        tapButton(app, "Multi")
        waitForText(app, "A: 0")

        // Tap counter B's "+" (the second of the three "+" buttons).
        let plusButtons = app.buttons.matching(NSPredicate(format: "label == %@", "+"))
        let second = plusButtons.element(boundBy: 1)
        XCTAssertTrue(second.waitForExistence(timeout: 10), "expected a second '+' button")
        second.tap()

        waitForText(app, "B: 1")
        // A and C untouched.
        waitForText(app, "A: 0")
        waitForText(app, "C: 0")
    }
}
